src/Fractionalizer.sol

```diff
@@ -0,0 +1,357 @@
+    using SafeERC20 for IERC20;
+
+    //todo: rename tokenId to ipnftId
+    event FractionsCreated(
+        uint256 indexed fractionId,
+        uint256 indexed tokenId,
+        address indexed tokenContract,
+        address emitter,
+        uint256 amount,
+        string agreementCid,
+        string name,
+        string symbol
+    );
+    event SalesActivated(uint256 fractionId, address paymentToken, uint256 paidPrice);
+    event TermsAccepted(uint256 indexed fractionId, address indexed signer, bytes signature);
+    event SharesClaimed(uint256 indexed fractionId, address indexed claimer, uint256 amount);
+    //listen for mints instead:
+    //event FractionsEmitted(uint256 fractionId, uint256 amount);
+
+    IPNFT ipnft;
+    SchmackoSwap schmackoSwap;
+
+    address feeReceiver;
+    uint256 fractionalizationPercentage;
+    mapping(uint256 => Fractionalized) public fractionalized;
+    mapping(address => mapping(uint256 => uint256)) claimAllowance;
+
+    address immutable tokenImplementation;
+
+    function initialize(IPNFT _ipnft, SchmackoSwap _schmackoSwap) public initializer {
+        __UUPSUpgradeable_init();
+        __Ownable_init();
+
+        ipnft = _ipnft;
+        schmackoSwap = _schmackoSwap;
+    }
+
+    modifier notClaiming(uint256 fractionId) {
+        if (address(fractionalized[fractionId].paymentToken) != address(0)) {
+            revert("already in claiming phase");
+        }
+        _;
+    }
+
+    constructor() {
+        tokenImplementation = address(new FractionalizedTokenUpgradeable());
+        _disableInitializers();
+    }
+
+    function setFeeReceiver(address _feeReceiver) public onlyOwner {
+        if (_feeReceiver == address(0)) {
+            revert ToZeroAddress();
+        }
+        feeReceiver = _feeReceiver;
+    }
+
+    function setReceiverPercentage(uint256 fractionalizationPercentage_) external onlyOwner {
+        fractionalizationPercentage = fractionalizationPercentage_;
+    }
+
+    function balanceOf(address owner, uint256 fractionId) public view returns (uint256) {
+        return fractionalized[fractionId].tokenContract.balanceOf(owner);
+    }
+
+    function totalSupply(uint256 fractionId) public view returns (uint256) {
+        return fractionalized[fractionId].tokenContract.totalSupply();
+    }
+
+    /**
+     * @notice
+     * @param ipnftId          uint256  the token id on the origin collection
+     * @param fractionsAmount  uint256  the initial amount of fractions issued
+     * @param agreementCid    bytes32  a content hash that identifies the terms underlying the issued fractions
+     */
+
+    function fractionalizeIpnft(uint256 ipnftId, uint256 fractionsAmount, string calldata agreementCid) external returns (uint256 fractionId) {
+        if (ipnft.totalSupply(ipnftId) != 1) {
+            revert("IPNFT supply must be 1");
+        }
+        if (ipnft.balanceOf(_msgSender(), ipnftId) != 1) {
+            revert("only owner can initialize fractions");
```

Comment:
can we use Custom Errors to save some gas

Reply:
sure, but they're far easier to write as long as you're in coding mode and they're quite hard to resolve on the client side. They mostly save gas during deployment, not during execution (only when the tx reverts), but I'm happily converting them, as we did on IPNFT, in the last roundtrip.